ros/src/driveworks/packages/autoware_driveworks_gmsl_interface/src/gmsl_camera_multiple/DrivesWorkApi.cpp

```diff
@@ -0,0 +1,491 @@
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ *  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ *  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ *  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ *  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ *  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+
+#include "DrivesWorkApi.hpp"
+
+namespace DrivesWork
+{
+	
+/*
+ * Constructor
+ *  This will initialise csi camera configuration and setting up sdk and sensors
+ */	
+DrivesWorkApi::DrivesWorkApi(DeviceArguments arguments):g_arguments(arguments)
+{
+  // init sdk, and start cameras
+  this->startCameras();
+}
+
+/*
+ * Destructor
+ */
+DrivesWorkApi::~DrivesWorkApi()
+{
+}
+
+/*
+ * Init SDK
+ */
+void DrivesWorkApi::initSdk(dwContextHandle_t *context)
+{
+  std::cout << "Init SDK .. " << std::endl;
+  // Instantiate Driveworks SDK context
+  dwContextParameters sdkParams;
+  memset(&sdkParams, 0, sizeof(dwContextParameters));
+  dwInitialize(context, DW_VERSION, &sdkParams);
+}
+
+/*
+ * Init SAL abstract layer
+ */
+void DrivesWorkApi::initSAL(dwSALHandle_t *sal, dwContextHandle_t context)
+{
+  std::cout << "Init SAL .. " << std::endl;
+  dwStatus result;
+  result = dwSAL_initialize(sal, context);
+  if (result != DW_SUCCESS) 
+  {
+	  std::cerr << "Cannot initialize SAL: " << dwGetStatusName(result) << std::endl;
+    exit(1);
+  }
+}
+
+
+/*
+ * Init sensors: csi ports,connected cameras, image width/height
+ */
+void DrivesWorkApi::initSensors(std::vector<Camera> *cameras, uint32_t *numCameras,
+                                dwSALHandle_t sal, DeviceArguments &arguments)
+{
+  std::cout << "Init Sensors .. " << std::endl;
+  std::string selector = arguments.get("selector_mask");
+  dwStatus result;
+  // identify active ports
+  int idx             = 0;
+  int cnt[3]          = {0, 0, 0};
+  std::string port[3] = {"ab", "cd", "ef"};
+  for (size_t i = 0; i < selector.length() && i < 12; i++, idx++) 
+  {
+	const char s = selector[i];
+	if (s == '1') 
+	{
+	  cnt[idx / 4]++;
+    }
+  }
+
+  // parsing arguments
+  (*numCameras) = 0;
+  for (size_t p = 0; p < 3; p++) 
+  {
+    if (cnt[p] > 0) 
+    {
+      std::string params;
+      params += std::string("csi-port=") + port[p];
+      params += ",camera-type=" + arguments.get((std::string("type_") + port[p]).c_str());
+      params += ",camera-count=4"; // when using the mask, just ask for all cameras, mask will select properly
+
+      if (selector.size() >= p*4) 
```

Comment:
Is this 4 for the four cameras in the GMSL port? If so, I think it'd better to use a variable or a constant for it.